doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,390 @@
+    {"name": "c2", "type": "int"}
+  ]
+}
+```
+
+Today, there is no way to instruct Materialize to include `doc` fields in the
+generated Avro schema. This poses a challenge for organizations with
+requirements on the use of documentation—e.g., a policy that requires that every
+field in every schema published to the organization's schema registry has
+documentation.
+
+## Goals
+
+Must haves:
+
+  * Enable users with organizational policies on the use of `doc` fields in Avro
+    schemas to use Avro-formatted sinks.
+
+Nice to haves:
+
+  * Consistency with Materialize's existing approach to generating Avro schemas
+    from SQL relations.
+
+  * Good ergonomics.
+
+## Non-goals
+
+None yet.
+
+## Detailed description
+
+### Design
+
+#### SQL
+
+We'll extend the `FORMAT AVRO` clause with a new `[KEY|VALUE] DOC ON
+[TYPE|COLUMN]` option. For example:
+
+```sql
+CREATE SINK ...
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ... (
+    DOC ON COLUMN <column-specifier> = '<documentation string>',
+    KEY DOC ON TYPE <type-specifier> = '<documentation string>',
+    VALUE DOC ON COLUMN <column-specifier> = '<documentation string>',
+    ...
+)
+WITH (...)
+```
+
+The option will be located in the existing parenthesized option list immediately
+following the existing `FORMAT AVRO` clause. After this change,
+`[KEY|VALUE] DOC ON [TYPE|COLUMN]` will live alongside the previously existing
+`AVRO KEY FULLNAME` and `AVRO VALUE FULLNAME` options and we expect to add more
+options to control the Avro schema in the future.
+
+Aside: the current design of the `USING CONFLUENT SCHEMA REGISTRY` option is
+historical and predates the introduction of the standardized option list. In the
+future, we may also move the `USING CONFLUENT SCHEMA REGISTRY` into the option
+list, where it can be freely intermixed with other options, e.g.:
+
+```sql
+FORMAT AVRO (
+    DOC ON ...,
+    CONFLUENT SCHEMA REGISTRY = '...',
+    AVRO KEY FULLNAME = '...',
+    KEY DOC ON ...,
+)
+```
+
+Each `[KEY|VALUE] DOC ON ...` option describes a documentation comment that should
+be attached to a single field in the Avro schema. A `KEY DOC ON ...` option
+applies only to the key schema; a `VALUE DOC ON ...` option applies only to the
+value schema; a `DOC ON ...` option applies to both types of schemas.
+
+Each `DOC ON` option applies to either a column or a type:
+
+  * Column comments (`[KEY|VALUE] DOC ON COLUMN ...`) name a column in the
+    format `[[[db.]schema.]object.]column`. If the object name is omitted, the
+    column name is taken to be a name of a column in the top level of the
+    materialized view. Object names are looked up according to usual SQL name
+    resolution rules for the search path and active database.
+
+  * Type comments (`[KEY|VALUE] DOC ON TYPE ...`) specify a type in the
+    format `[[db.]schema.]object`. Object names are looked up according to usual
+    SQL name resolution rules for the search path and active database.
```

Comment:
Do we have a motivating use case for this feature? It seems like a convenience when folks don't want to specify field-level doc comments, but I'm not sure in what case it is useful and I'm not aware of prior art in eg. other tools that specify avro schemas.

Reply:
Gah! Sorry. Good catch. I totally botched this while switching from Avro-centric field specifiers to SQL-centric field specifiers. Can you take a look at the new text and see if it makes more sense? AIUI Avro allows the `doc` attribute in _two_ places: on record types themselves, and on each _field_ of a record type (which may itself be a record type with another doc coumment). So `DOC ON TYPE` is necessary to get the `doc` attribute on the record types, and `DOC ON COLUMN` is necessary to get the `doc` attribute onto the record fields.

@moulimukherjee can you PTAL too? I took the liberty of just pushing up a fix because it was easier to explain by fixing the design doc than typing it out as a comment!

Reply:
Yep, the update makes sense to me - thanks!

Reply:
lgtm too, thanks for the update!